### DISCUSSION TITLE
How to avoid using for loop in these kind of  situations

### DISCUSSION THREAD

**Discussion by ZhenxuanZhang-Jensen** [scikit-hep/awkward#1500] [Q&A] [answered]
Hi experts,
   Currently, I am trying to use uproot to load a root file, and change it as a numpy array and do the cut, and calculate the cut efficiency. I keep this in mind for a long time, and I'm always thinking about this kind of situation which I met a lots of times. Can we don't use for loop, but use awkward array to do the exact same thing? such as creating an awkward array to save all the efficiency and then using slice to do the cut?
Best regards,
Zhenxuan
```sig_xgb_roc = []
bkg_xgb_roc = []
for i in np.linspace(-1,1,1000):
    sig_xgb_roc.append(event_signal_xgb_weights[event_signal_xgb['DiphotonMVA_self'].array(library = 'np') > i].sum() / event_signal_xgb_weights.sum())
    bkg_xgb_roc.append((event_bkg_pp_xgb_weights[event_bkg_pp_xgb['DiphotonMVA_self'].array(library = 'np') > i].sum() +  event_bkg_pfff_xgb_weights[event_bkg_pfff_xgb['DiphotonMVA_self'].array(library = 'np') > i].sum()) / ( event_bkg_pp_xgb_weights.sum() + event_bkg_pfff_xgb_weights.sum())) ```

**Comment by jpivarski** [comment, marked as answer]
Well, first thing: the [uproot.TBranch.array](https://uproot.readthedocs.io/en/latest/uproot.behaviors.TBranch.TBranch.html#array) method reads data and you don't want to read and re-read the same data over and over. You could pass a `cache` to the `array` method so that it would check the `cache` for an already-read value instead of getting it from the file, but it's probably easier to just move the

```python
signal_xgb = event_signal_xgb['DiphotonMVA_self'].array(library = 'np')
bkg_pp_xgb = event_bkg_pp_xgb['DiphotonMVA_self'].array(library = 'np')
bkg_pfff_xgb = event_bkg_pfff_xgb['DiphotonMVA_self'].array(library = 'np')
```

read commands out of the loop and just use those arrays, rather than reading them, in the loop.

Secondly, you can remove the sums that don't depend on `i` outside the loop on `i`, so that they don't get recomputed each time. That is,

```python
signal_xgb_weights_sum = event_signal_xgb_weights.sum()
bkg_pp_xgb_weights_sum = event_bkg_pp_xgb_weights.sum()
bkg_pfff_xgb_weights_sum = event_bkg_pfff_xgb_weights.sum()
```

What remains is a for loop over 1000 values of `i`, and generally pure Python is acceptable for loops of only 1000. That is, the fact that you have to step over Python statements 1000 times is not going to be a pain point: the Python overhead will be milliseconds. _But_, what you have in that loop is a few sums, and if those sums are over a large amount of data, then doing 1000 of those might be painful.

So first, try the fixes described above and if it's still a problem, you might want to try using [np.sort](https://numpy.org/doc/stable/reference/generated/numpy.sort.html), [np.cumsum](https://numpy.org/doc/stable/reference/generated/numpy.cumsum.html) and/or [np.searchsorted](https://numpy.org/doc/stable/reference/generated/numpy.searchsorted.html). You want to add up everything in an array that's greater than a given `i`. Instead of repeatedly cutting and summing, sort the data so that all the elements greater than `i` would be to the right of `i`, do one cumulative sum, so that the sum of everything greater than `i` is the value at `i`, and that might be fast enough already because 1000 tests would just be 1000 array-lookups (reading values from the cumulative sum 1000 times). If it _really, really_ matters after that, you can even vectorize the lookups with `searchsorted`, but I doubt it will matter—unless this example with 1000 was a small-scale test and you'll actually be doing it with millions or billions of `i` values.

All while you are making these incremental improvements, check the results of the optimized code against the results you already have with the slow code. Don't try to do all the improvements at once—small steps with correctness-checking at each step (i.e. "test driven development")—will likely involve less debugging.

**Reply by ZhenxuanZhang-Jensen** [reply]
Thanks a lot, Jim. The advices and suggestions for you are always very helpful, I do learn a lot, and indeed, it helps me to reduce the time consumption. Actually, I am finding a universal way to avoid using loop. Maybe the example I gave is a too easy one, let me give you another problem I met before, and maybe I can touch on the essence of the array usages this time.

I used to train a DNN model to choose the correct 4 signal jets over bunches of jets(different events have different numbers, maybe like 7jets) of each event. To mark the signal and bkgs in training, I choose to use arrays to save 7 jets in each event, and the arrays in which the first 4 jets are signal jets will be marked as signals.like:[1,2,3,4,5,6,7],[2,1,3,4,5,6,7] -> mark as signal,(equal to 1 in training),[5,1,2,3,4,6,7],[6,1,2,3,4,5,7]-> mark as bkg,(equal to 0 in training)

 So the point is, we will have 7! combinations of each event, and 4! of them will be marked as signals and other will be marked as bkgs. And it's all good by now, we can get the model. And the difficult part is in the evaluation(application), as you know, for each event we have 7! combination, and I need to give them all a score then choose the best score combination, and the first four jets in that best score combination of each event is are the signal jets we are looking for. So if we use for loop to do these, we need to loop over all the events, and in each event, we need to loop over 7! combination to find the best score array, and get the information of the first four indexes of that array and save that in the tree. And it seems that we can't save it all to a big array to do the model prediction since it only can do it one by one. So for me, it seems very difficult to decrease the time consumption or use a more advanced way to improve these. And I think it's very common for everyone if we need to use DNN and do the evaluation, I don't know if there are any better ways to cut time or maybe awkward array can better handle this situation. 
Best,
Zhenxuan

**Reply by jpivarski** [reply]
> And it seems that we can't save it all to a big array to do the model prediction since it only can do it one by one.

7! = 5040 is a lot of combinations to keep in memory as an intermediate array. When you do `ak.combinations(jets, 7)` in Awkward, it makes an IndexedArray of the `jets`, so it's not making full copies of each jet, but still that's a lot of integers to have in memory. You'll want to iterate over them, not store them in memory.

That leads me to think you'll want to use Numba for this problem, with `for` loops, but compiled. Or, actually, RDataFrame.

```python
>>> import awkward as ak    # or import awkward._v2 as ak
>>> import numba as nb
>>> jets = ak.Array([[{"px": 1}, {"px": 2}], [], [{"px": 3}, {"px": 4}, {"px": 5}]])
>>> @nb.njit
... def walk_over_combinations(jets):
...     for event_jets in jets:
...         for i in range(len(event_jets)):
...             for j in range(i + 1, len(event_jets)):
...                 print(event_jets[i].px, event_jets[j].px)
... 
>>> walk_over_combinations(jets)
1 2
3 4
3 5
4 5
```

Or, with very recent versions of Awkward Array (from GitHub's `main` branch),

```python
>>> import awkward._v2 as ak   # has to be version 2
>>> jets = ak.Array([[{"px": 1}, {"px": 2}], [], [{"px": 3}, {"px": 4}, {"px": 5}]])
>>> rdf = ak.to_rdataframe({"jets": jets})
>>> rdf2 = rdf.Define("total", "double total = 0; for (auto jet : jets) { total += jet.px(); } return total")
>>> rdf2.AsNumpy(["total"])
{'total': ndarray([ 3.,  0., 12.])}
```

and presumably you can put more exciting C++ code in the `Define`. Record objects like individual jets are on-demand classes with their fields as methods, and lists are iterable things (use `auto` for an instance type—it's not a guessable type name). If the list only contains numbers, they're [ROOT::VecOps::RVecs](https://root.cern/doc/master/classROOT_1_1VecOps_1_1RVec.html). We don't have a working `ak.from_rdataframe` yet, though.

Both of these methods will allow you to compile loops over your very large number of combinations. At the moment, I can't think of any other way to approach this problem.

Perhaps [this would have been a good tool](https://github.com/jpivarski/PartiQL), but I never finished it.

**Reply by ZhenxuanZhang-Jensen** [reply]
I have tried numba before and it seems I can't use TensorFlow in numba, so I can't do the prediction in numba, so as the Rdataframe.

**Reply by agoose77** [reply]
I don't fully understand what you're doing, but at a high level, you wouldn't use Numba / RDataFrame to perform your model fitting. Instead, you'd use it to manipulate/transform your arrays, treating it like a black box:

```mermaid
flowchart LR
data[Data Source] --> select[Jagged Selection] --> tensor[TensorFlow]
```

**Reply by ZhenxuanZhang-Jensen** [reply]
yes, What I want to do is to use Numba or RDataFrame to accelerate the DNN model evaluation and get the score of each given dataset that needs to be produced in the for loop. So I need to predict the score one by one in the for loop inside Numba/RDataFrame which means I need to use tensorflow or more specific `model.predict(array)` inside Numba/RDataFrame. But I tried, and it seems that doesn't work, so it just gets me back to the very beginning of my problem, how to accelerate the whole running process...

**Reply by jpivarski** [reply]
The trick is to use these tools to build up what you're calling "`array`" here. That's what would use Numba or RDataFrame, the formatting of that data. Numba and RDataFrame don't recognize TensorFlow objects or functions as types (it would have to have an explicit Numba or C++ interface; it's possible there's a C++ interface, but I doubt you want to use it).

The `model.predict` function surely takes large, multi-event datasets as possible inputs, perhaps as a two-dimensional NumPy array.

If your TensorFlow model takes jagged data, the interface it probably wants is [RaggedTensor](https://www.tensorflow.org/guide/ragged_tensor). We don't have an explicit function to convert an Awkward Array into a RaggedTensor (issue #1466), but it wouldn't be hard to do that conversion, zero-copy. [RaggedTensor.from_row_lengths](https://www.tensorflow.org/api_docs/python/tf/RaggedTensor#from_row_lengths) needs the values that [ak.num](https://awkward-array.readthedocs.io/en/latest/_auto/ak.num.html) provides (high-level route), and [RaggedTensor.from_row_splits](https://www.tensorflow.org/api_docs/python/tf/RaggedTensor#from_row_splits) takes the `offsets` of a `ListOffsetArray` (low-level route, fewer computations). If you go the low-level route, you can find the `ListOffsetArray` inside an array's `layout`, and if your jagged array is based on a `ListArray` or if the `offsets` doesn't start with `0` (which TensorFlow needs), there's a `toListOffsetArray64(start_at_zero=True)` method that you can call on it.

**Reply by ZhenxuanZhang-Jensen** [reply]
Okay. Thanks for all the answers, I think I know where I need to take a look now.
